File: 백준 9663 N-Queen -백트랙킹-/복습/Main.java
import java.util.Scanner;

public class Main {
    static int n;
    static int result = 0;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        for (int i = 1; i <= n; i++) {
            int[] num = new int[n + 1];
            num[1] = i;
            dfs(num, 1);
        }
        System.out.println(result);
    }

    static void dfs(int[] num, int row) {
        if (row == n) {
            result++;
            return;
        }
        for (int j = 1; j <= n; j++) {
            num[row + 1] = j;
            if (check(num, row + 1)) {
                dfs(num, row + 1);
            }
        }
    }

    static boolean check(int[] num, int row) {
        for (int i = 1; i < row; i++) {
            if (num[i] == num[row]) { //같은열존재
                return false;
            }
            if (Math.abs(i - row) == Math.abs(num[i] - num[row])) { //대각선존재
                return false;
            }
        }
        return true;
    }
}

File: 백준 9663 N-Queen -백트랙킹-/복습/Main.kt
import java.util.*

internal var n: Int = 0
internal var result = 0

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    n = sc.nextInt()
    for (i in 1..n) {
        val num = IntArray(n + 1)
        num[1] = i
        dfs(num, 1)
    }
    println(result)
}

internal fun dfs(num: IntArray, row: Int) {
    if (row == n) {
        result++
        return
    }
    for (j in 1..n) {
        num[row + 1] = j
        if (check(num, row + 1)) {
            dfs(num, row + 1)
        }
    }
}

internal fun check(num: IntArray, row: Int): Boolean {
    for (i in 1 until row) {
        if (num[i] == num[row]) { //같은열존재
            return false
        }
        if (Math.abs(i - row) == Math.abs(num[i] - num[row])) { //대각선존재
            return false
        }
    }
    return true
}